dll/cpl/hotplug/enum.c

```diff
@@ -4,4 +4,292 @@
+    {
+        ClassGuid = GUID_DEVCLASS_UNKNOWN;
+    }
+
+    /* Get the image for the class this device is in */
+    SetupDiGetClassImageIndex(&pHotplugData->ImageListData,
+                              &ClassGuid,
+                              pImageIndex);
+}
+
+static
+HTREEITEM
+InsertDeviceTreeItem(
+    _In_ HTREEITEM hParent,
+    _In_ DEVINST DevInst,
+    _In_ PHOTPLUG_DATA pHotplugData)
+{
+    WCHAR szDisplayName[MAX_DEVICE_DISPLAYNAME_LEN];
+    INT nClassImage;
+    TVINSERTSTRUCTW tvItem;
+
+    GetDeviceDisplayInfo(DevInst,
+                         pHotplugData,
+                         szDisplayName,
+                         ARRAYSIZE(szDisplayName),
+                         &nClassImage);
+
+    ZeroMemory(&tvItem, sizeof(tvItem));
+    tvItem.hParent = hParent;
+    tvItem.hInsertAfter = TVI_LAST;
+
+    tvItem.item.mask = TVIF_STATE | TVIF_TEXT | TVIF_PARAM | TVIF_IMAGE | TVIF_SELECTEDIMAGE;
+    tvItem.item.state = TVIS_EXPANDED;
+    tvItem.item.stateMask = TVIS_EXPANDED;
+    tvItem.item.pszText = szDisplayName;
+    tvItem.item.iImage = nClassImage;
+    tvItem.item.iSelectedImage = nClassImage;
+    tvItem.item.lParam = (LPARAM)DevInst;
+
+    return TreeView_InsertItem(pHotplugData->hwndDeviceTree, &tvItem);
+}
+
+static
+VOID
+DevTreeRecursiveInsertSubDevices(
+    _In_ HTREEITEM hParentItem,
+    _In_ DEVINST ParentDevInst,
+    _In_ PHOTPLUG_DATA pHotplugData)
+{
+    HTREEITEM hTreeItem;
+    DEVINST ChildDevInst;
+    CONFIGRET cr;
+
+    //DPRINT("DevTreeRecursiveInsertSubDevices()\n");
+
+    cr = CM_Get_Child(&ChildDevInst, ParentDevInst, 0);
+    if (cr != CR_SUCCESS)
+    {
+        //DPRINT("No child! %lu\n", cr);
+        return;
+    }
+
+    hTreeItem = InsertDeviceTreeItem(hParentItem,
+                                     ChildDevInst,
+                                     pHotplugData);
+    if (hTreeItem != NULL)
+    {
+        DevTreeRecursiveInsertSubDevices(hTreeItem,
+                                         ChildDevInst,
+                                         pHotplugData);
+    }
+
+    for (;;)
+    {
+        cr = CM_Get_Sibling(&ChildDevInst, ChildDevInst, 0);
+        if (cr != CR_SUCCESS)
+        {
+            //DPRINT("No sibling! %lu\n", cr);
+            return;
+        }
+
+        hTreeItem = InsertDeviceTreeItem(hParentItem,
+                                         ChildDevInst,
+                                         pHotplugData);
+        if (hTreeItem != NULL)
+        {
+            DevTreeRecursiveInsertSubDevices(hTreeItem,
+                                             ChildDevInst,
+                                             pHotplugData);
+        }
+    }
+}
+
+VOID
+EnumHotpluggedDevices(
+    _In_ PHOTPLUG_DATA pHotplugData)
+{
+    SP_DEVINFO_DATA did = { 0 };
+    HDEVINFO hdev;
+    int idev;
+    DWORD dwCapabilities, dwSize;
+    ULONG ulStatus, ulProblem;
+    HTREEITEM hTreeItem;
+    CONFIGRET cr;
+
+    //DPRINT1("EnumHotpluggedDevices()\n");
```

Comment:
If not needed anymore (and iirc. there are other such commented DPRINTs in the hotplug), they may be removed.

Reply:
OK.